controller/lib/hal/hal.h

```diff
@@ -289,11 +289,23 @@ class HalApi {
       IntEnable();
   }
 
+  // Reset the hardware peripheral in STM32
+  void crc32_reset();
+  // Accumulate CRC32 in STM32 hardware
+  void crc32_accumulate(uint8_t d);
+  // Get CRC32 accumulated in STM32 hardware
+  uint32_t crc32_get();
+  // Calculate CRC32 for data buffer
+  uint32_t crc32(uint8_t *data, uint32_t len);
```

Comment:
Looks good, except I think we need native implementations of these functions?  Otherwise we can't call these functions in code that gets compiled for platform native.  That would mean we can't unit test comms.

Reply:
Yes, sure, didn't have a chance to work on it more.

Reply:
But in general I would fix it when it breaks - now no tests are braking, when I will be rolling out framing then they will break.

Side thought - can we pass HalApi as a constructor parameter to all its clients? This would get rid of dichotomy of HalApi with all the #define noise. We could then have HalHw and HalForTesting classes. Hal being global is Arduino framework legacy we are dragging really. For that matter we could even split HalCRC, HalGpio, HalSerial, etc

Reply:
> Side thought - can we pass HalApi as a constructor parameter to all its clients? This would get rid of dichotomy of HalApi with all the #define noise. We could then have HalHw and HalForTesting classes.

To do this we'd need a HalApi "interface" class, right?

```
class HalApi {
  virtual void doThing() = 0;
};
class HalHw : public HalApi {
  virtual void doThing() override { ... }
};

class Controller {
  explicit Controller(HalApi& hal) : hal_(hal) {}

  HalApi& hal_;
};

```

The ifdefs are definitely ugly.  I did it this way to avoid having virtual functions everywhere.  Maybe that's a premature optimization -- in theory, LTO can devirtualize, but in practice I am not sure it's going to be 100% effective, nor how we'd be able to tell.

> Hal being global is Arduino framework legacy we are dragging really.

I guess I'm not as sure about that.  Like, there are no circumstances under which it would be ok to have two HAL instances alive at the same time (right?).

Reply:
Ah, the whole c++ inheritance machine comes to play when you use virtual functions and that might be expensive on small MCU, right? We can measure at least the time it takes to call a function by flipping a pin.

Reply:
> Ah, the whole c++ inheritance machine comes to play

Yeah, but it's not any more than you'd get in C.  That is, the cost should be identical to something like the following C code.

```
struct HalApiVtable {
  void (*doFoo)(HalApi*, int x, int y);
  int (*doBar)(HalApi*, int x, int y);
};
struct HwHal {
  HalApiVtable* vtable_;
  HalApi hal_api_members_;
  int whatever_;
};
```

where the implementation of one of these "virtual functions" does something like the following.

```
void doFoo_ForHwHal(HalApi* base, int x, int y) {
  HwHal* this = reinterpret_cast<HwHal*>(reinterpret_cast<char*>(base) - sizeof(HalApiVtable*));
  return this->whatever_ + x - y;
}
```

Notably this means you can't inline anything, unless LTO saves our butts.